tests/scanner/scanners/retention_scanner_test.py

```diff
@@ -0,0 +1,251 @@
+    data_creater.AddLefecycleDict("Delete", 99, None, None, None, None)
+    res.append(data_creater.get_resource())
+
+    data_creater = frsd.FakeBucketDataCreater('p1bkt22', p1)
+    data_creater.AddLefecycleDict("Delete", 100, None, None, None, None)
+    res.append(data_creater.get_resource())
+
+    data_creater = frsd.FakeBucketDataCreater('p1bkt23', p1)
+    data_creater.AddLefecycleDict("Delete", 101, None, None, None, None)
+    res.append(data_creater.get_resource())
+
+
+    data_creater = frsd.FakeBucketDataCreater('p2bkt31', p2)
+    data_creater.AddLefecycleDict("Delete", 149, None, None, None, None)
+    res.append(data_creater.get_resource())
+
+    data_creater = frsd.FakeBucketDataCreater('p2bkt32', p2)
+    data_creater.AddLefecycleDict("Delete", 150, None, None, None, None)
+    res.append(data_creater.get_resource())
+
+    data_creater = frsd.FakeBucketDataCreater('p2bkt33', p2)
+    data_creater.AddLefecycleDict("Delete", 300, None, None, None, None)
+    res.append(data_creater.get_resource())
+
+    data_creater = frsd.FakeBucketDataCreater('p2bkt34', p2)
+    data_creater.AddLefecycleDict("Delete", 450, None, None, None, None)
+    res.append(data_creater.get_resource())
+
+    data_creater = frsd.FakeBucketDataCreater('p2bkt35', p2)
+    data_creater.AddLefecycleDict("Delete", 451, None, None, None, None)
+    res.append(data_creater.get_resource())
+
+    data_creater = frsd.FakeBucketDataCreater('p2bkt4', p2)
+    data_creater.AddLefecycleDict("Delete", 600, None, None, None, None)
+    res.append(data_creater.get_resource())
+
+    data_creater = frsd.FakeBucketDataCreater('p2bkt5', p2)
+    data_creater.AddLefecycleDict("Delete", 200, None, None, None, None)
+    res.append(data_creater.get_resource())
+
+    global _fake_bucket_list_01
+    _fake_bucket_list_01 = res
+    return res
+
+
+def ModifyBucketLifeCycle(proj_res, bucketdata, action, age, created_before, matches_storage_class, num_newer_versions, is_live):
+
+    Resource = collections.namedtuple(
+        'Resource',
+        # fields based on required fields from Resource in dao.py.
+        ['full_name', 'type', 'name', 'parent_type_name', 'parent',
+            'data'],
+    )
+    datajson = json.loads(bucketdata[4])
+    datajson["lifecycle"]["rule"] = []
+    datajson["lifecycle"]["rule"].append(GetLefecycleDict(action, age, created_before, matches_storage_class, num_newer_versions, is_live))
+    return Resource(full_name=bucketdata[0],type=bucketdata[1],parent_type_name=bucketdata[2],name=bucketdata[3],parent=proj_res,data=json.dumps(datajson))
+
+
+class RetentionScannerTest(ForsetiTestCase):
+
+    def setUp(self):
+        """Set up."""
+
+
+    #"lifecycle": {"rule": [{"action": {"type": "Delete"}, "condition": {"age": 20, "numNewerVersions": 5}}, {"action": {"type": "Delete"}, "condition": {"age": 10, "numNewerVersions": 99}}]},
+    @unittest.skipIf(do_not_test_old_cases, 'debug new test cases')
+    def test_scanner_01(self):
```

Comment:
You can follow the pattern of other `*_scanner_test.py` files, where the scanner_test mocks out the rule engine and focuses more on the retrieve, find_violations and output_results methods. The rule_engine test focuses more on the logic of finding violations in resources.